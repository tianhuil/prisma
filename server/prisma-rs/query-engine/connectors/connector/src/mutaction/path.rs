use crate::filter::{Filter, NodeSelector};
use prisma_models::prelude::*;

pub struct Path {
    segments: Vec<PathSegment>,
}

pub enum PathSegment {
    ToOne(RelationFieldRef),
    ToMany(RelationFieldRef, NodeSelector),
    ToManyFilter(RelationFieldRef, Option<Filter>),
}

impl From<RelationFieldRef> for PathSegment {
    fn from(rf: RelationFieldRef) -> PathSegment {
        PathSegment::ToOne(rf)
    }
}

impl From<(RelationFieldRef, NodeSelector)> for PathSegment {
    fn from(t: (RelationFieldRef, NodeSelector)) -> PathSegment {
        PathSegment::ToMany(t.0, t.1)
    }
}

impl From<(RelationFieldRef, Option<Filter>)> for PathSegment {
    fn from(t: (RelationFieldRef, Option<Filter>)) -> PathSegment {
        PathSegment::ToManyFilter(t.0, t.1)
    }
}

impl Path {
    pub fn append(&mut self, segment: PathSegment) {
        self.segments.push(segment);
    }
}
