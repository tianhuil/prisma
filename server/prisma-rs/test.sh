#!/bin/bash
set -e

git submodule update --init || true
cargo test -- --test-threads 1