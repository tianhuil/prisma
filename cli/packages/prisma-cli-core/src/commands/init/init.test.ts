// import * as nock from 'nock'
import { mocki } from './inquirer-mock-prompt'
import { Config } from 'prisma-cli-engine'
import Init from './init'
import { writeToStdIn, DOWN, ENTER } from '../../test/writeToStdin'
import { getTmpDir } from '../../test/getTmpDir'
import * as fs from 'fs-extra'
import * as path from 'path'
import * as cuid from 'scuid'

function makeMockConfig(mockInquirer?: any) {
  const home = getTmpDir()
  const cwd = path.join(home, cuid())
  fs.mkdirpSync(cwd)
  return {
    config: new Config({ mock: true, home, cwd, mockInquirer }),
    home,
    cwd,
  }
}

function getFolderContent(folder) {
  return fs
    .readdirSync(folder)
    .map(f => {
      let file = fs.readFileSync(path.join(folder, f), 'utf-8')
      if (f === 'docker-compose.yml') {
        file = normalizeDockerCompose(file)
      }
      return { [f]: file }
    })
    .reduce((acc, curr) => ({ ...acc, ...curr }), {})
}

async function testChoices(choices) {
  const mockInquirer = mocki(choices)
  const { config, home, cwd } = makeMockConfig(mockInquirer)
  const result = await Init.mock({ mockConfig: config })

  expect(getFolderContent(cwd)).toMatchSnapshot()
  expect(makeOutputDeterministic(result.out.stdout.output)).toMatchSnapshot()
  expect(makeOutputDeterministic(result.out.stderr.output)).toMatchSnapshot()
}

function makeOutputDeterministic(output: string) {
  return output
    .split('\n')
    .filter(l => !l.includes('Connecting to database'))
    .join('\n')
}

describe.skip('init', () => {
  test('choose local', async () => {
    await testChoices({
      choice: 'local',
    })
  })

  test('test project', async () => {
    await testChoices({
      choice: 'Use existing database',
      dbType: 'postgres',
      host: 'localhost',
      port: 5432,
      database: process.env.TEST_PG_DB,
      user: process.env.TEST_PG_USER || 'postgres',
      password: process.env.TEST_PG_PASSWORD,
      schema: undefined,
      alreadyData: false,
    })
  })
})

export function normalizeDockerCompose(dc) {
  return dc
    .split('\n')
    .filter(
      l =>
        !l.trim().startsWith('user:') &&
        !l.trim().startsWith('password:') &&
        !l.trim().startsWith('database:') &&
        !l.trim().startsWith('image:'),
    )
    .join('\n')
}
