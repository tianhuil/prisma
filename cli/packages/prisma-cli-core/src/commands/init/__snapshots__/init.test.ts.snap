// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init choose local 1`] = `
Object {
  "datamodel.prisma": "type User {
  id: ID! @unique
  name: String!
}
",
  "docker-compose.yml": "version: '3'
services:
  prisma:
    restart: always
    ports:
    - \\"4466:4466\\"
    environment:
      PRISMA_CONFIG: |
        port: 4466
        # uncomment the next line and provide the env var PRISMA_MANAGEMENT_API_SECRET=my-secret to activate cluster security
        # managementApiSecret: my-secret
        databases:
          default:
            connector: postgres
            host: postgres
            rawAccess: true
            port: 5432
            migrations: true
  postgres:
    restart: always
    environment:
      POSTGRES_USER: prisma
      POSTGRES_PASSWORD: prisma
    volumes:
      - postgres:/var/lib/postgresql/data
volumes:
  postgres:
",
  "prisma.yml": "endpoint: http://localhost:4466
datamodel: datamodel.prisma",
}
`;

exports[`init choose local 2`] = `
"
Created 3 new files:                                                                          

  prisma.yml           Prisma service definition
  datamodel.prisma    GraphQL SDL-based datamodel (foundation for database)
  docker-compose.yml   Docker configuration file

Next steps:

  1. Start your Prisma server: docker-compose up -d
  2. Deploy your Prisma service: prisma deploy
  3. Read more about Prisma server:
     http://bit.ly/prisma-server-overview
"
`;

exports[`init choose local 3`] = `""`;

exports[`init test project 1`] = `Object {}`;

exports[`init test project 2`] = `""`;

exports[`init test project 3`] = `""`;
